velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp

```diff
@@ -0,0 +1,144 @@
+    scale_ = requestedType_->asShortDecimal().scale();
+  } else {
+    scale_ = requestedType_->asLongDecimal().scale();
+  }
+  version_ = convertRleVersion(stripe.getEncoding(encodingKey).kind());
+  auto data = encodingKey.forKind(proto::Stream_Kind_DATA);
+  valueDecoder_ = createDirectDecoder</*isSigned*/ true>(
+      stripe.getStream(data, params.streamLabels().label(), true),
+      stripe.getUseVInts(data),
+      sizeof(DataT));
+
+  // [NOTICE] DWRF's NANO_DATA has the same enum value as ORC's SECONDARY
+  auto secondary = encodingKey.forKind(proto::Stream_Kind_NANO_DATA);
+  scaleDecoder_ = createRleDecoder</*isSigned*/ true>(
+      stripe.getStream(secondary, params.streamLabels().label(), true),
+      version_,
+      memoryPool_,
+      stripe.getUseVInts(secondary),
+      LONG_BYTE_SIZE);
+}
+
+template <typename DataT>
+uint64_t SelectiveDecimalColumnReader<DataT>::skip(uint64_t numValues) {
+  numValues = SelectiveColumnReader::skip(numValues);
+  valueDecoder_->skip(numValues);
+  scaleDecoder_->skip(numValues);
+  return numValues;
+}
+
+template <typename DataT>
+void SelectiveDecimalColumnReader<DataT>::seekToRowGroup(uint32_t index) {
+  auto positionsProvider = formatData_->seekToRowGroup(index);
+  valueDecoder_->seekToRowGroup(positionsProvider);
+  scaleDecoder_->seekToRowGroup(positionsProvider);
+  VELOX_CHECK(!positionsProvider.hasNext());
+}
+
+template <typename DataT>
+template <bool dense>
+void SelectiveDecimalColumnReader<DataT>::readHelper(RowSet rows) {
+  vector_size_t numRows = rows.back() + 1;
+  ExtractToReader extractValues(this);
+  common::AlwaysTrue filter;
+  DirectRleColumnVisitor<
+      int64_t,
+      common::AlwaysTrue,
+      decltype(extractValues),
+      dense>
+      visitor(filter, this, rows, extractValues);
+
+  // decode scale stream
+  if (version_ == velox::dwrf::RleVersion_1) {
+    decodeWithVisitor<RleDecoderV1<true>>(scaleDecoder_.get(), visitor);
+  } else {
+    decodeWithVisitor<RleDecoderV2<true>>(scaleDecoder_.get(), visitor);
+  }
+
+  // copy scales into scaleBuffer_
+  ensureCapacity<int64_t>(scaleBuffer_, numValues_, &memoryPool_);
+  scaleBuffer_->setSize(numValues_ * sizeof(int64_t));
+  memcpy(
+      scaleBuffer_->asMutable<char>(),
+      rawValues_,
+      numValues_ * sizeof(int64_t));
+
+  // reset numValues_ before reading values
+  numValues_ = 0;
+  valueSize_ = sizeof(DataT);
+  ensureValuesCapacity<DataT>(numRows);
+
+  // decode value stream
+  facebook::velox::dwio::common::
+      ColumnVisitor<DataT, common::AlwaysTrue, decltype(extractValues), dense>
+          valueVisitor(filter, this, rows, extractValues);
+  decodeWithVisitor<DirectDecoder<true>>(valueDecoder_.get(), valueVisitor);
+  readOffset_ += numRows;
+}
+
+template <typename DataT>
+void SelectiveDecimalColumnReader<DataT>::read(
+    vector_size_t offset,
+    RowSet rows,
+    const uint64_t* incomingNulls) {
+  VELOX_CHECK(!scanSpec_->filter());
```

Comment:
Will we support filters?

Reply:
Can we do this through another PR？since Timestamp Column Reader does’t support this too, we can implement it together.